Fix premature note cleanup on error

julianrother opened the pull request:
### Component/Part
Backend (realtime)

### Description

This PR fixes that `interruptConnection()` cleans up note state without first checking if other clients are still connected to the note.

The call to `interruptConnection()` in case of a permission error causes the behavior described in #3894. Since `failConnection()` is also called in this case and since it performs the same cleanup as `interruptConnection()`, but with properly checking for other clients, I removed the call to `interruptConnection()`.

It took me some time to understand why the other call to `interruptConnection()` was added in the first place, whether/when it is executed and if it could cause the same problem as the other call:

```js
function finishConnection (socket, noteId, socketId) {
  // if no valid info provided will drop the client
  if (!socket || !notes[noteId] || !users[socketId]) {
    return interruptConnection(socket, noteId, socketId)
  }
  // ...
```

`!socket` can never be true otherwise the function call `finishConnection(socket, noteId, socket.id)` would have already crashed. `!notes[noteId]` can never be true, because right before the call to `finishConnection()`, `notes[noteId]` is either populated or it is estabilshed that `notes[noteId]` is already populated.

However, if the client is the first client for a note, `startConnection()` fetches the note object asynchronously. If the client disconnects before the promise is resolved, `disconnect()` would have cleaned up `users[socketId]` before the call to `finishConnection()` and the third condition would be true. This also matches the [commit that introduced this check](https://github.com/hedgedoc/hedgedoc/commit/d3a23ad72f1cdde4e7b87e8d79c678b27d98b8c2) ("Fixed realtime.js finishConnection user might be undefined issue"). So in this case cleanup of the `notes` global and `clearSocketQueue()`/`connectNextSocket()` are neccesary.

Anyway, I don't think this call to `interruptConnection()` (and the unconditional `delete notes[noteId]`) does any harm, since it only happens if the client is the first client for a note, but I couldn't resist refactoring it.

### Steps

<!-- Please tick all steps this PR performs (if something is not necessary, please remove it) -->

- [x] Added implementation
- [x] Added changelog snippet
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### Related Issue(s)
Fixes #3894, probably fixes #1992

mrdrogdrog reviewed (approved):
Thanks for your contribution. It looks good and works.

Also congratulation for surviving the HedgeDoc 1.x spaghetti code. :yum:

davidmehren commented on public/docs/release-notes.md:
```diff
@@ -7,6 +7,9 @@
 - Add dedicated healthcheck endpoint at /_health that is less resource intensive than /status.
 
+### Bugfixes
+- Fix that permission errors can break existing connections to a pad, cause inconsistent pad content and changes not being saved
```
```suggestion
- Fix that permission errors can break existing connections to a note, causing inconsistent note content and changes not being saved
```

davidmehren reviewed (changes requested):
Just nitpicking on the changelog entry, everything else looks very sensible 👍🏻

mrdrogdrog reviewed (changes requested):
Please squash the docs commit into the other one.

julianrother commented:
Thank you for the feedback! I've now squashed the docs commit and rebased everything onto the current master, so feel free to merge :)

Do you have plans for a 1.x release with this fix?

mrdrogdrog commented:
Maybe today. A new 1.x release is overdue.

davidmehren reviewed (approved):
Thank you for contributing!
